File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.cc
#include <iostream>
#include <queue>
using namespace std;

const int MX = 1002;
string board[MX];
int n, m;

int dx[4] = {1,0,-1,0};
int dy[4] = {0,1,0,-1};

int dist1[MX][MX];
bool ch1[MX][MX];
int dist2[MX][MX];
bool ch2[MX][MX];

void bfs(pair<int,int> s, int dist[][MX], bool ch[][MX]){
    queue<pair<int,int>> q;
    q.push(s);
    ch[s.first][s.second] = true;
    
 
    for(int i = 0; i < MX;i++){
        for(int j = 0; j<MX;j++){
            dist[i][j] = 1e8;
        }
    }
    
    int d = 0;
    while(!q.empty()){
        int sz = q.size();
        while(sz--){
            int x = q.front().first;
            int y = q.front().second;
            q.pop();
            
            dist[x][y] = d;
            for(int i = 0; i < 4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if(nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if(ch[nx][ny]) continue;
                if(board[nx][ny] == '1'){
                    dist[nx][ny] = d + 1;
                    ch[nx][ny] = 1;
                    continue;
                }
                
                q.push({nx,ny});
                ch[nx][ny] = true;
            }
        }
        d++;
    }
}

int main()
{
    cin >> n >> m;
    
    for(int i =0; i<n;i++){
        cin >> board[i];
    }
    
    bfs({0,0}, dist1, ch1);
    bfs({n-1,m-1},dist2,ch2);
    
    int mn = 1e9;
    
    for(int i =0; i <n;i++){
        for(int j = 0; j <m;j++){
            mn = min(dist1[i][j] + dist2[i][j], mn);
        }
    }
    
    if(mn >= 1e8) cout << -1;
    else cout << mn + 1;
    
    
    return 0;
}

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static int N, M;
	static int map[][] = new int[1002][1002];
	static final int dx[] = new int[] { 1, 0, -1, 0 };
	static final int dy[] = new int[] { 0, -1, 0, 1 };
	static int dist1[][] = new int[1002][1002];
	static int dist2[][] = new int[1002][1002];
	static final int INF = 1000000;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		for (int i = 0; i < N; i++) {
			String str = br.readLine();
			for (int j = 0; j < M; j++) {
				map[i][j] = str.charAt(j) - '0';
			}
		}

		bfs(0, 0, dist1);
		bfs(N - 1, M - 1, dist2);

		int mn = INF;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				mn = Math.min(mn, dist1[i][j] + dist2[i][j]);
			}
		}

		if (mn >= INF) {
			System.out.println(-1);
		} else {
			System.out.println(mn + 1);
		}
	}

	static void bfs(int sx, int sy, int[][] dist) {
		Queue<int[]> q = new ArrayDeque<>();
		boolean ch[][] = new boolean[N][M];

		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				dist[i][j] = INF;
			}
		}
		q.add(new int[] { sx, sy });
		ch[sx][sy] = true;
		dist[sx][sy] = 0;
		while (!q.isEmpty()) {
			int[] cur = q.poll();
			int x = cur[0];
			int y = cur[1];

			for (int dir = 0; dir < 4; dir++) {
				int nx = x + dx[dir];
				int ny = y + dy[dir];
				if (nx < 0 || nx >= N || ny < 0 || ny >= M)
					continue;
				if (ch[nx][ny])
					continue;
				if (map[nx][ny] == 1) {
					dist[nx][ny] = Math.min(dist[nx][ny], dist[x][y] + 1);
					continue;
				}
				q.add(new int[] { nx, ny });
				ch[nx][ny] = true;
				dist[nx][ny] = dist[x][y] + 1;

			}
		}
	}

}